### ISSUE TITLE
this makes me sad :(

### ISSUE THREAD

**Issue by spheenik** [yasp-dota/rapier#4]
but you're a free man, of course...

**Comment by albertcui** [comment]
Haha honestly, I believe this was mostly a pet project by @howardchung. Rapier is really slow compared to clarity, so I don't think we'd ever switch :) Thanks for all of your work!

**Comment by spheenik** [comment]
those words help to dry the tears...

Well, from a technical side I'm interested in how fast you can get this.
After looking over the code, it's still going to be quite some way to go until you can read packet entities.

and regarding this: https://github.com/yasp-dota/rapier/blob/master/BitStream.js#L14-L73
Optimizing this one to every last iota you can find is of utmost importance :)

**Comment by howardchung** [comment]
A JavaScript parser is necessary if we want to do something with in-browser parsing.  I'm not sure yet if we are going in that direction but I figured it would be nice to release a library for others to use if they wanted.  It's also educational to implement a replay parsing library :)

Performance wise so far it is taking about 17 seconds on a 50 minute replay without entities.  We will likely continue using clarity in production.

Latest work is in entities branch.  Main steps left are implementing the Huffman tree and decoder methods.

**Comment by spheenik** [comment]
Do you know this? 

http://d2playback.com

It's alice compiled with emscripten to work in the browser.

**Comment by howardchung** [comment]
Yeah, although:
- Alice is not updated to work with source2. @invokr seems to be working on Manta.
- This lets developers work in JS rather than c++, and avoids the build step required to compile with emscripten.

**Comment by spheenik** [comment]
I've worked with him on decoding some of the stuff, and the latest I hear is that he is going to make a version of alice supporting Source 2. There's a performance crown to defend against those pesky java developers!

And yes, I agree that working with emscripten output might be hard, and having a javascript library is better. It's just that it's a bad tool for the job. It's hard to see how you're ever going to get it to a speed where a user would accept it running in the browser. 17 seconds, and you didn't even parse a single baseline.... It will take tens of minutes in the end.

Well, for me it seems like plainly unfeasible to do it in javascript. Kind of like trying to break the 100m world record against Ursain Bolt in a  badly oiled wheelchair...

**Comment by howardchung** [comment]
Well, http://devilesk.com/dota2/apps/replay/viewer/ is also alice compiled to JS with emscripten.  It takes about 30 seconds to parse, and uses entities (since it can get hero, creep, and ward positions).

The emscripten output is still just JavaScript, so it should be possible to complete a parse w/entities in approximately that timeframe using Rapier.  Still about 2x slower than clarity, but not unreasonable.

The 17 seconds can be cut in half by skipping GameEvents (mostly combat log).  For some reason it seems really expensive to parse those packets... although it may be just because there are something like 40,000 combat log events in the replay I'm testing

**Comment by spheenik** [comment]
Clarity is able to do a parse with entities in under 3 secs. So it's more like 10x slower.

**Comment by howardchung** [comment]
Probably most of the time is spent assembling the messages and sending it to output then.  I'm going off the times that it takes to complete a parse for YASP purposes (which includes writing output).

Since the 30 seconds figure also (presumably) includes the time to create and process the output, it's probably inflated over the pure parse time.

**Comment by invokr** [comment]
Regarding Alice 2:
There will definietly be an Alice version for Source 2 and I'll make sure it compiles down to js. I'm currently working with dotabuff on Manta, I'll work on Alice once that is finished, which is probably sometime this week.

On Topic:
I don't know devilesk code, but the performance penalty for using Alice in JS was about 50% for me, so still < 3 seconds. He is parsing the whole replay and re-structuring the data, I don't think that would run much faster on the desktop. The audio code that I use for d2playback only runs 30% slower in Firefox, no kidding.

You will never ever in a million years even come close to that with normal javascript. The code that emscripten outputs takes advantage of a whole number of compiler optimization passes that would render normal JS code unreadable and unmaintainable. I remember distinct benchmarks between emscripten and javascript libraries, and em was ahead of them every time by a large margin. 

Right now, with emscripten being renamed / specified to WebAssembly, a lot of languages will add support for it. Even some of the more academic languages, like Idris, provide a javascript backend. Go also supports javascript via gopher js.

WIth that being said, I don't want to tell you how to run your project. Pure Javascript, however, is probably the worst language for the job.

**Comment by spheenik** [comment]
Right. If you do output, it'll take longer.
I also agree everybody should embrace the language he wants.
I would be impressed if you get this to a state where it's faster that anything emscripten.

**Comment by howardchung** [comment]
Thanks for the feedback, guys.  You have way more experience at this than I do :)

Based on your comments, there is no way this library will achieve high enough performance for production YASP.  This project is largely academic, and I want it to get to at least a usable state for developers who are willing to accept a large performance penalty in exchange for working fully in JavaScript.

I may play around with emscripten once alice2 is complete to achieve end-to-end JS for production YASP.

**Comment by spheenik** [comment]
A big problem is the lacking bitstream performance. It's implementation is the basis of all things speed. 
Also snappy.js, as well as the protobuf deserializing needs to be as fast as it can get.
Maybe you could use emscripten to compile the native snappy and protobuf implementations and use those for a more readable implementation of the rest?

**Comment by spheenik** [comment]
And out of interest: I see you forked dota2-replay-js to get a base for rapier.
It looks capable of parsing entities (for a source 1 replay).
How long does it take?

**Comment by howardchung** [comment]
I think any bitwise operations tend to be slower in JS due to the Number->int->Number conversion (since everything in JS is a double), and that slowness adds up when we're doing millions of them per parse.
- BitStream.  Based off Manta's implementation (clarity's seems similar).  Read the relevant bytes from the backing buffer, then shift them around to extract the bits needed.  Not sure it can be improved significantly while staying in pure JS.
- Snappy.  I did try the native version (node-snappy is a binding to a native implementation).  Parse times were about two seconds faster.  The binding doesn't work in the browser, so I decided to just use the pure JS version.  It's probably possible to write it in C++ and emscripten that single function into JS for something that works in the browser and node.
- Protobufs.  This might be the cause of some serious slowdown.  I am using the protobufjs library, which is used by node-steam.  It works great there, but it's not under the heavy load that replay parsing puts it under.  Similarly to snappy there is a library with a native binding (protobuf), but again that will require a swap between node and browser (where we either need to use protobufjs or something custom compiled w/emscripten).

So basically to improve performance I'd just have to write all the "hot" sections of the code in C++ and then compile them into emscriptened JS.  That way the library can still offer the same JS API to developers who want to work in JS while maintaining acceptable performance.  Considering this is just a hacky side project, I probably won't do that now, but maybe in the future if a lot of people decide to start using it...

dota2-replay-js was never able to completely parse a replay (errors all over the place), so I just borrowed a few functions from it and decided to write rapier instead.

I'm not quite sure what the fairest way to benchmark a parser is anyway.  Is it simply to listen for all messages (like clarity's @OnMessage or Manta's OnAny) and simply do nothing?

**Comment by spheenik** [comment]
> So basically to improve performance I'd just have to write all the "hot" sections of the code in C++ and > then compile them into emscriptened JS. 

That was my thought. Probably @invokr can drop a line on whether that's possible at all.

When I benchmark the entity parser, I just let it process all PacketEntities messages, but output nothing.
To be able to parse packet entities, you need the "instancebaseline" string table, as well as the send tables.

Simply doing `@OnMessage`is slower, since it'll get any and all packets.

**Comment by spheenik** [comment]
and also the guy who decided that every number in JS is a double should be prevented to ever get near a computer again!

**Comment by howardchung** [comment]
For a comprehensive test shouldn't a benchmark cover everything?  (building string tables, entities, all packets)  Otherwise you're only covering part of what's in a replay.

**Comment by invokr** [comment]
The graph on the Alice page parses every packet, all the entities and baselines, though does zero work with the data. I think that's the fairest way to benchmark. Three seconds is still pretty fast though, my earliest version was around 10 seconds I think.

Re Hotcode:
It's possible to do it that way, though I'd probably consider the whole parser "hotcode". Copying the data between the EM-Memory and JS might create some unnecessary overhead in the long run.
Everything you want to call in JS needs to be exported as a C function, so exporting OnAny and some interface for entities might be less painful than exporting every utility library.

**Comment by spheenik** [comment]
> For a comprehensive test shouldn't a benchmark cover everything?

Not if I only want to test a specific usecase.
For example I might only be interested in extracting the combat log.
Then why touch entities at all?

**Comment by howardchung** [comment]
But shouldn't you assume the user is interested in everything, benchmark based on that, and quote a worst-case time which can be improved by skipping packets?

@invokr 's benchmark sounds pretty good to me.

**Comment by spheenik** [comment]
No you should not assume the user is interested in everything.
For example voice data. Who's gonna need that?

**Comment by howardchung** [comment]
d2playback does?

But that's why it's a worst-case number.  You could give a few more "real-world" cases where you drop some of the most esoteric packets.

In the interest of easy comparison between parsers though, it's just simplest to compare every library's time to parse every packet+entities.

**Comment by invokr** [comment]
Unless you actually do decoding, the packets alone shouldn't add to much overhead. Benchmarking the decoding would probably zero-out results among parsers because everyone is using the same libcelt implementation anyways. Protobuf uses a zero-copy solution internally so the largest portion of the voice packet is just a pointer to memory that's already loaded anyway.

I agree with @howardchung's notion that the easiest way to compare parsers based on benchmarks would be to process all available data. This doesn't 100% reflect real-world scenarios, but hardly any benchmark ever does.

I do care about the performance, though anything that parses a replay in less than 5 secs is fine in my book. Actually doing something with the data adds more time anyway.

**Comment by spheenik** [comment]
do I have to process string table updates or just read the protobuf?
do I have to unpack the usermessages for S1 replays?

**Comment by invokr** [comment]
The alice benchmark does process all stringtables, doesn't unpack usermessages though.

I'll see that I write a simple benchmark suite for the S2 version that parses (and unpacks) all available data.
